six/modules/c++/samples/test_large_offset.cpp

```diff
@@ -0,0 +1,164 @@
+* This file is part of six-c++
+* ==========================================================================
+*
+* (C) Copyright 2004 - 2017, MDA Information Systems LLC
+*
+* six-c++ is free software; you can redistribute it and/or modify
+* it under the terms of the GNU Lesser General Public License as published by
+* the Free Software Foundation; either version 3 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public
+* License along with this program; If not,
+* see <http://www.gnu.org/licenses/>.
+*
+*/
+
+#include <cstring>
+
+#include <io/TempFile.h>
+#include <mem/SharedPtr.h>
+
+#include <six/Container.h>
+#include <six/NITFReadControl.h>
+#include <six/NITFWriteControl.h>
+#include <six/sicd/ComplexXMLControl.h>
+#include <six/sicd/Utilities.h>
+#include <six/sidd/DerivedXMLControl.h>
+#include <six/sidd/Utilities.h>
+
+namespace
+{
+static const size_t NUM_ROWS_IN_FIRST_SEGMENT = 47673; // By manual inspection
+static const size_t ROWS_TO_SKIP = NUM_ROWS_IN_FIRST_SEGMENT - 1000;
+
+void createNITF(const std::string& outputPathname,
+        const six::DataType& datatype)
+{
+    six::XMLControlRegistry registry;
+    registry.addCreator(six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<
+                    six::sicd::ComplexXMLControl>());
+    registry.addCreator(six::DataType::DERIVED,
+            new six::XMLControlCreatorT<
+                    six::sidd::DerivedXMLControl>());
+    const std::vector<std::string> schemaPaths;
+
+    std::auto_ptr<six::Data> data;
+    if (datatype == six::DataType::COMPLEX)
+    {
+        data = six::sicd::Utilities::createFakeComplexData();
+        data->setNumRows(52440); // (11GB / 4) ^ (1/2)
+        data->setNumCols(52440);
+        data->setPixelType(six::PixelType::RE16I_IM16I);
+    }
+    else
+    {
+        data = six::sidd::Utilities::createFakeDerivedData();
+        data->setNumRows(74152); // (11GB / 2) ^ (1/2)
+        data->setNumCols(74162);
+        data->setPixelType(six::PixelType::MONO16I);
+    }
+
+    const size_t bytesPerRow = data->getNumCols() * data->getNumBytesPerPixel();
+    const size_t imageSize = data->getNumRows() * bytesPerRow;
+
+    mem::SharedPtr<six::Container> container(
+            new six::Container(data->getDataType()));
+    container->addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.setXMLControlRegistry(&registry);
+    writer.initialize(container);
+
+    mem::ScopedArray<six::UByte> imageData(new six::UByte[imageSize]);
+    for (size_t ii = 0; ii < imageSize; ++ii)
+    {
+        imageData[ii] = static_cast<six::UByte>(ii);
```

Comment:
This is definitely better, but we could still get unlucky and not catch a bug if our underlying reading logic was goofed up.  Since this is being treated as ubyte data, there are effectively only 256 unique values.  We could be reading from the totally wrong spot in the file and just happen to land on the same repeating values (granted, you only have a 1 / 256 chance of that).

To make really sure, how about changing the SICD check over to use `RE32F_IM32F` and then in this loop use incrementing floating point values (I assume you'll then need to cut the number of rows you're using in half to get the segment boundary math to work out)?  That way you'll be unique all the way through the image (might as well set the real and imaginary components to unique values too just to be sure... maybe one positive and one negative or something).  If you treat the SIDD as 16-bit values, that'll give you 2^16 rather than 2^8 unique values which isn't perfect but is good enough since we'll have an exhaustive SICD test.  Obviously you'll need a cast and different check down at the `checkNITF()` code that's in sync with this too.